### ISSUE TITLE
Array(0.0, -0.0).distinct result is wrong.

### ISSUE THREAD

**Issue by WeichenXu123** [scala/bug#11995]
## reproduction steps

See comment https://github.com/scala/bug/issues/11994#issuecomment-627070712
0.0 and -0.0 are different,

then why Array(0.0, -0.0).distinct get result only contains 0.0 ?
using Scala (2.12),
```scala
scala> Array(0.0, -0.0).distinct
res1: Array[Double] = Array(0.0)

```

but sometimes, 0.0 and -0.0 are both kept after distinct:
~~~scala
import scala.util.Random
val rng = new Random(3)

val a1 = Array.tabulate(200)(_=>rng.nextDouble * 2.0 - 1.0) ++ Array.fill(20)(0.0) ++ Array.fill(20)(-0.0)
a1.distinct.sorted.foreach(x => print(x.toString + "\n"))
~~~

will get output
```
...
-0.009292684662246975
-0.0033280686465135823
-0.0
0.0
0.0022219556032221366
0.02217419561977274
...
```

## problem

 Array(0.0, -0.0).distinct result is wrong.

**Comment by WeichenXu123** [comment]
@sjrd Could you explain this ?

**Comment by SethTisue** [comment]
please use https://users.scala-lang.org for questions

**Comment by SethTisue** [comment]
this was already fixed in Scala 2.13.0

**Comment by SethTisue** [comment]
how it got fixed, I don't know. here's a more minimal reproduction:

```
scala 2.12.11> (Array.tabulate(55)(1 + _ * 0.001) :+ 0.0 :+ -0.0).distinct.filter(_ == 0.0)
res104: Array[Double] = Array(0.0)

scala 2.12.11> (Array.tabulate(56)(1 + _ * 0.001) :+ 0.0 :+ -0.0).distinct.filter(_ == 0.0)
res105: Array[Double] = Array(0.0, -0.0)
```

what's magic about the number 56, why does it cross over at that point? some weirdness with hash buckets? 🤷‍♂️

**Comment by SethTisue** [comment]
I've closed the ticket because the problem doesn't seem to exist in 2.13, but if someone gets to the bottom of why this is happening in 2.12.x, perhaps that same someone, or a different someone, would also like to submit a fix to our 2.12.x branch

**Comment by Jasper-M** [comment]
[This commit}(https://github.com/scala/scala/commit/3a6f637b33056df80b93e743fe274f623925ea21) might have something to do with it.

> - Hash set computation is fixed to properly implement Scala’s equality
  semantics for numeric types.
